keps/sig-autoscaling/20191111-autoscaling-with-multiple-thresholds.md

```diff
@@ -0,0 +1,202 @@
+---
+title: Multi thresholds based Autoscaling.
+authors:
+  - "@CharlyF"
+owning-sig: sig-autoscaling
+participating-sigs:
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-11-11
+last-updated: 2019-11-11
+status: provisional
+see-also:
+  - "/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md"
+replaces:
+superseded-by:
+---
+
+# Multi thresholds based Autoscaling
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories-optional)
+    - [Canonical use case of the bounds to control the primary scaling condition.](#Canonical-use-case-of-the-bounds-to-control-the-primary-scaling-condition)
+    - [Restriction of the scaling events.](#Restriction-of-the-scaling-events)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [The Pseudo Code](#the-pseudo-code)
+- [Design Details](#design-details)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+<!-- /toc -->
+
+## Summary
+
+This is a proposal to extend the Horizontal Pod Autoscaler (or HPA) Controller to allow for a better configurability of the scaling behavior.
+Namely, avoiding the sawtooth pattern by introducing two thresholds as low and high bounds.
+We (Datadog) have been using a Custom Resource internally, the Watermark Pod Autoscaler or WPA, and we believe the community could benefit from this feature. 
+Feel free to check out [the official repository](https://github.com/DataDog/watermarkpodautoscaler) and give us your feedback.
+
+Worth mentioning, we have also included some of the features introduced by @gliush
+and @arjunrn in in the [Configurable Scale Velocity KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md) in our controller as we believe those features are must-haves for the next verison of the HPA Controller.
+
+
+## Motivation
+
+- Set a high and a low bound at the resource level to better control autoscaling events.
+
+### Goals
+
+The goal is to improve the ability to customize the parameters triggering a scaling event. We identified 4 vectors of improvement:
+
+- Specifying multiple thresholds on a per HPA basis.
```

Comment:
I don't think that we need both tolerance and high/low watermark thresholds.  For example, these two configurations are equivalent:

```yaml
    metrics:
    - resource:
        name: cpu
        highWatermark: 60
        lowWatermark: 40
      type: Resource
    tolerance: 0.0
```

```yaml
    metrics:
    - resource:
        name: cpu
        targetAverageUtilization: 50
      type: Resource
    tolerance: 0.2
```

Reply:
Yes, you are absolutely right - I considered doing this at first, but from my experience and from discussing with other autoscaling users, it seems like people think in terms of boundaries first. So they know that a webserver should not handle more than X rq/s or that below Y rq/s they have headroom and can downscale. With the threshold logic, it requires a little bit more maths (think about the boundaries and then find the middle and set the right percentage).

This is why I took this route in the first place, for the user experience - But from an implementation perspective and especially given the complexity of the controller, I am in favour of revamping this KEP to simply move the threshold to the HPA level from the overarching controller.

Reply:
@josephburnett Both of them are equivalent but consider what the second example would like when the user wants to set the high watermark to `65` and low water mark to `40`. Then the target has to be set to `52.5` and the tolerance to `0.238`. It's much more straightforward for the users to specify upper and lower bounds. IMO tolerance is still somewhat an internal detail of the controller. It's only exposed through the tolerance flag. It's not too late to move away from that model for the API.
@CharlyF I am of the same opinion that users tend to think more in terms of upper and lower bounds than tolerances. I would recommend that don't just implement tolerances in the API but the original proposal of upper and lower water marks. It's far more useful. Of course the tolerance should be ignored when the watermarks are specified.

Reply:
Thanks! I don't want to block on this, I am going to keep this solution in the `Alternative solution` section, to keep the discussion opened.